File: imgfun.py
import numpy as np
def Img2Mat(img):
    """Reshapes an RGB image with dimension(r, c, 3) to a matrix with dimension [r*c, 3].
    
    Each row of the matrix corresponds to a pixel in row-major order."""
    return np.reshape(img, (img.shape[0]*img.shape[1], 3))

def Mat2Img(m, r, c):
    """Inverse of Img2Matrix, with number of rows and columns specified."""
    return np.reshape(m, (r,c,3))

def AssignMeans(means, pmat):
    """Compute nearest means

    Args:
        means: K x d
        pmat: N x d
    
    Returns:
        N x 1 vector with indices of nearest means
    """
    K = means.shape[0]
    N = pmat.shape[0]
    kdists = np.zeros((K, N))
    for i in range(0, K):
        print "computing distances to %dth mean" % i
        kdists[i, :] = np.sqrt(np.sum(np.square(np.tile(means[i, :], (N,1)) - pmat), axis=1))
    print "Compute argmin"
    a = np.argmin(kdists, axis=0)
    print a.shape
    return a

def ComputeMeans(clusters, pmat, K):
    """Compute means given cluster assignments
    
    Args:
        clusters: N x 1 cluster assignments for pmat, has entries from 0..K-1
        pmat: N x d
        K: number of clusters
        
    Returns:
        K cluster centers, K x d, where i's center is in row i-1"""
    print "Start computing means"
    [N, d] = pmat.shape
    means = np.zeros((K, d))
    for i in range(0, K):
        print "Computing mean %d" % i
        print clusters.shape
        members = np.reshape(clusters == i, (N, 1))
        print members.shape
        num_points = np.sum(members)
        print num_points
        if num_points > 0:
            means[i, :] = np.sum(np.multiply(np.tile(members, (1, d)),
                                 pmat), axis=0) / num_points
    return means

def Kmeans(pmat, K, mini, maxi):
    """Run K means on data
    
    The K means are initialized uniformly in [mini, maxi]
    
    Args:
        K: number of means
        pmat: data with N data points and d dimensions, N x d numpy array
        
    Returns:
        means: K x d numpy array representing the K means
        clusters: N x 1, cluster assignments, indexed from 0 to K-1"""
    means = np.random.rand(K, 3) * (maxi - mini) + mini
    print pmat.shape
    [N, d] = pmat.shape
    clusters = np.zeros((N))
    prev_clusters = np.zeros((N))
    iterations = 0
    while True:
        iterations += 1
        print "Running iteration %d" % iterations
        clusters = AssignMeans(means, pmat)
        print clusters
        print clusters.shape
        print prev_clusters.shape
        agreements = np.sum(np.equal(prev_clusters, clusters))
        print agreements
        if agreements == N:
            # This ends because K-means is guaranteed to converge
            print "Converged"
            break
        print "before compute means"
        means = ComputeMeans(clusters, pmat, K)
        prev_clusters = np.copy(clusters)
    return [means, clusters]
    
def GenKmeansImage(img, K):
    """Generates an rgb-image by performing K-means on img
    
    Args:
        img: numpy image
        K: number of clusters
    Returns:
        New image"""
    imat = Img2Mat(img)
    N = imat.shape[0]
    [means, clusters] = Kmeans(imat, K, 0, 255)
    newimg = np.zeros((N, 3))
    for i in range(N):
        newimg[i, :] = means[clusters[i], :]
    return Mat2Img(newimg, img.shape[0], img.shape[1]).astype("uint8")

File: imgfun.lua
local P = {}
imgfun = P

function P.img2mat(img)
    return torch.reshape(img, 3, img:size()[2]*img:size()[3]):t()
end

function P.mat2img(mat, h, w)
    return torch.reshape(mat:t(), 3, h,w)
end

function P.assign_means(means, mat)
    -- means: K x d
    -- mat -- N x d
    -- returns N x 1 assignments
    local K = means:size()[1]
    local N = mat:size()[1]
    local kdists = torch.zeros(K, N)
    for i = 1,K do
        kdists[{i, {}}] = (means[{i, {}}]:repeatTensor(N,1)-mat):pow(2):sum(2):sqrt()
    end
    local mins, indices = kdists:min(1)
    return indices:t()      
end

function P.compute_means(clusters, mat, K)
    -- clusters: N x 1 cluster assignments for pmat, has entries from 1..K
    -- mat: N x d
    -- K: Number of clusters
    local N = mat:size()[1]
    local d = mat:size()[2]
    local means = torch.zeros(K, d)
    for i=1,K do
        local members = clusters:eq(i):reshape(N, 1)
        local num_points = members:sum()
        if num_points > 0 then
            means[{i, {}}] = members:repeatTensor(1, d):double():cmul(mat):sum(1) / num_points
        end
    end
    return means
end

function P.kmeans(mat, K, mini, maxi)
    local means = torch.randn(K, 3) * (maxi - mini) + mini
    local N = mat:size()[1]
    local d = mat:size()[2]

    local prev_clusters = torch.LongTensor(N, 1)
    local clusters = prev_clusters:clone()
    local iters = 0
    while true do
        iters = iters + 1
        clusters = P.assign_means(means, mat)
        local agreements = torch.eq(prev_clusters, clusters):sum()
        if agreements == N then
            break
        end
        means = P.compute_means(clusters, mat, K)
        prev_clusters:set(clusters)
    end
    print("Kmeans iterations: ", iters)
    return means, clusters
end

function P.gen_kmeans_image(img, K)
    local imat = P.img2mat(img)
    local N = imat:size()[1]
    local m, clusters = P.kmeans(imat, K, 0, 1)
    local img2 = torch.zeros(N, 3)
    local x1 = m[{clusters[1], {}}]
    local y1 = img2[{1, {}}]
    for i=1,N do
        img2[{i, {}}] = m[{clusters[i][1], {}}]
    end
    return P.mat2img(img2, img:size()[2], img:size()[3])
end

return P
